src/library/compiler/ir_interpreter.cpp

```diff
@@ -0,0 +1,759 @@
+                // originally borrowed parameters because the wrapper will decrement these after the call.
+                // Basically the wrapper is more homogeneous (removing both boxed and borrowed parameters) than we
+                // would need in this instance.
+                for (size_t i = 0; i < args.size(); i++) {
+                    if (param_borrow(decl_params(d)[i])) {
+                        inc(m_arg_stack[old_size + i]);
+                    }
+                }
+            }
+            // HACK: `curry` wants a closure object instead of just a function pointer
+            object * cls = alloc_closure(e.m_addr, 2, 1);
+            r = curry(cls, args.size(), &m_arg_stack[old_size]);
+            free_heap_obj(cls);
+        } else {
+            if (decl_tag(d) == decl_kind::Extern) {
+                throw exception(sstream() << "unexpected external declaration '" << fn << "'");
+            }
+            r = eval_body(decl_fun_body(d));
+        }
+        pop_frame(r);
+        return r;
+    }
+public:
+    explicit interpreter(environment const & env) : m_env(env) {}
+
+    uint32 run_main(int argc, char * argv[]) {
+        decl d = get_fdecl("main");
+        array_ref<param> const & params = decl_params(d);
+        if (params.size() == 2) { // List String -> IO UInt32
+            lean_object * in = lean_box(0);
+            int i = argc;
+            while (i > 1) {
+                i--;
+                lean_object * n = lean_alloc_ctor(1, 2, 0);
+                lean_ctor_set(n, 0, lean_mk_string(argv[i]));
+                lean_ctor_set(n, 1, in);
+                in = n;
+            }
+            m_arg_stack.push_back(in);
+        } else { // IO UInt32
+            lean_assert(params.size() == 1);
+        }
+        object * w = io_mk_world();
+        m_arg_stack.push_back(w);
+        push_frame("main", 0);
+        w = eval_body(decl_fun_body(d));
+        pop_frame(w);
+        if (io_result_is_ok(w)) {
+            // NOTE: in an awesome hack, `IO Unit` works just as well because `pure 0` and `pure ()` use the same
+            // representation
+            int ret = unbox(io_result_get_value(w));
+            dec_ref(w);
+            return ret;
+        } else {
+            io_result_show_error(w);
+            dec_ref(w);
+            return 1;
+        }
+    }
+
+    // closure stub
+    object * stub_m(object ** args) {
+        decl d(args[1]);
+        size_t old_size = m_arg_stack.size();
+        for (size_t i = 0; i < decl_params(d).size(); i++) {
+            m_arg_stack.push_back(args[16 + i]);
+        }
+        push_frame(decl_fun_id(d), old_size);
+        object * r = eval_body(decl_fun_body(d));
+        pop_frame(r);
+        return r;
+    }
+
+    // closure stub stub
+    static object * stub_m_aux(object ** args) {
+        interpreter * self = reinterpret_cast<interpreter *>(unbox(args[0]));
```

Comment:
I am wondering if this approach works in the following scenario:
- We create an `interpreter` `interp`, and execute some Lean IR code.
- A closure is created and is stored in some global `IO.Ref` created using compiled code.
- `interp` is deleted.
The closure in the `IO.Ref` is now referencing dead code.

Reply:
Yes, `IO.Ref`s are very problematic with *any* interactive usage that may lead to rewinding and reexecuting code.

Reply:
`IO.Ref` is not essential for the issue I have described above. Replace it with `Environment`, and it is still the same problem. The abstract issue is:
- In the code above, it seems closures contain a reference to the `interpreter` object.
- Thus, they are only valid while the `interpreter` object remains "alive". 
- This seems problematic since we can create an `interpreter`, execute some code using it, store the resulting closure in an `Environment` object, destroy the `interpreter`, and then try to access the closure stored in the `Environment`.

Reply:
This should be fixed with the multi-threading change. The only interpreter data inside the closure will be the `Environment`. If a thread does not have an active interpreter, it will be created on call, regardless of whether that is due to multi-threading or otherwise delayed closures.

Reply:
I added the simplest-possible support for multi-threading (https://github.com/leanprover/lean4/pull/33/commits/190c7ae49f66e65849d40bbc134c56c9a901e815).
Different interpreters only share the `Environment`, all stacks and caches are always fresh. For stacks and the symbol cache this should be fine I believe; `dlsym` is reasonably fast, using only ~10% CPU even without any caching. Not sure about the global constants, should we have a thread-shared cache for them?